src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -97,169 +135,335 @@ func newCommitLogSource(opts Options, inspection fs.Inspection) bootstrap.Source
-	switch strategy {
-	case bootstrap.BootstrapSequential:
-		return true
-	}
-	return false
+func (s *commitLogSource) AvailableData(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) (result.ShardTimeRanges, error) {
+	return s.availability(ns, shardsTimeRanges, runOpts)
 }
 
-func (s *commitLogSource) AvailableData(
+func (s *commitLogSource) AvailableIndex(
 	ns namespace.Metadata,
 	shardsTimeRanges result.ShardTimeRanges,
 	runOpts bootstrap.RunOptions,
 ) (result.ShardTimeRanges, error) {
 	return s.availability(ns, shardsTimeRanges, runOpts)
 }
 
-// ReadData will read all commitlog files on disk, as well as as the latest snapshot for
+type readNamespaceResult struct {
+	namespace               bootstrap.Namespace
+	dataMetadata            bootstrap.NamespaceResultDataMetadata
+	indexMetadata           bootstrap.NamespaceResultIndexMetadata
+	dataAndIndexShardRanges result.ShardTimeRanges
+}
+
+// Read will read all commitlog files on disk, as well as as the latest snapshot for
 // each shard/block combination (if it exists) and merge them.
 // TODO(rartoul): Make this take the SnapshotMetadata files into account to reduce the
 // number of commitlogs / snapshots that we need to read.
-func (s *commitLogSource) ReadData(
-	ns namespace.Metadata,
-	shardsTimeRanges result.ShardTimeRanges,
-	runOpts bootstrap.RunOptions,
-) (result.DataBootstrapResult, error) {
-	if shardsTimeRanges.IsEmpty() {
-		return result.NewDataBootstrapResult(), nil
+func (s *commitLogSource) Read(
+	namespaces bootstrap.Namespaces,
+) (bootstrap.NamespaceResults, error) {
+	timeRangesEmpty := true
+	for _, elem := range namespaces.Namespaces.Iter() {
+		namespace := elem.Value()
+		dataRangesNotEmpty := !namespace.DataRunOptions.ShardTimeRanges.IsEmpty()
+		indexRangesNotEmpty := namespace.Metadata.Options().IndexOptions().Enabled() &&
+			!namespace.IndexRunOptions.ShardTimeRanges.IsEmpty()
+		if dataRangesNotEmpty || indexRangesNotEmpty {
+			timeRangesEmpty = false
+			break
+		}
+	}
+	if timeRangesEmpty {
+		// Return empty result with no unfulfilled ranges.
+		return bootstrap.NewNamespaceResults(namespaces), nil
 	}
 
 	var (
 		// Emit bootstrapping gauge for duration of ReadData
-		doneReadingData        = s.metrics.data.emitBootstrapping()
+		doneReadingData        = s.metrics.emitBootstrapping()
 		encounteredCorruptData = false
 		fsOpts                 = s.opts.CommitLogOptions().FilesystemOptions()
 		filePathPrefix         = fsOpts.FilePathPrefix()
+		namespaceResults       = make(map[string]*readNamespaceResult)
+		initialTopologyState   *topology.StateSnapshot
 	)
 	defer doneReadingData()
 
-	// Determine which snapshot files are available.
-	snapshotFilesByShard, err := s.snapshotFilesByShard(
-		ns.ID(), filePathPrefix, shardsTimeRanges)
-	if err != nil {
-		return nil, err
+	startSnapshotsRead := s.nowFn()
+	s.log.Info("read snapshots start")
```

Comment:
nit: add time to log?

Reply:
Each log always has a timestamp - I've also added the duration to the `read snapshots end` with `took` var.